lib/utils/README.md

```diff
@@ -2,19 +2,375 @@
+Combined with [aggregate initialization](https://en.cppreference.com/w/cpp/language/aggregate_initialization), we are able to construct a Person as follows:
+```cpp
+Person p = { "donald", "knuth", 85, true };
+```
+and any subset of the fields would raise an error at compile time.
+
 ### Limitations
 
+`visitable` types have two primary limitations. First, they do not support initialization with `(...)`:
+```cpp
+Person p{ "donald", "knuth", 85, true }; // CORRECT
+Person p2("robert", "tarjan", 75, true); // ERROR
+```
+Secondly, template types cannot be visitable (we hope to remove this limitation in the distant future), but instantiations of them can.
+```cpp
+template <typename T>
+struct MyLists {
+  std::vector<T> list1;
+  req<std::vector<T>> list2;
+};
+FF_VISITABLE_STRUCT(MyLists, list1, list2); // ERROR
+
+using MyInts = MyLists<int>;
+
+FF_VISITABLE_STRUCT(MyInts, list1, list2); // CORRECT
+```
+A smaller limitation is that `FF_VISITABLE_STRUCT` only works from within the `FlexFlow` namespace (this is not much of an issue as all of the `FlexFlow` code resides in a single namespace).
+
+### Advanced Features
+
+While `FF_VISITABLE_STRUCT` matches the behavior of many product types in FlexFlow's codebase, there are exceptions. Many of these resemble the code below:
+```cpp
+struct Cow { ... };
+
+struct TownPopulation {
+  std::vector<Person> people;
+  std::vector<Cow> cows;
+};
+```
+Unlike in the `Person` example, `TownPopulation` has an obvious default value: an empty town (i.e., both people and cow are empty).
+However, if we write
+```cpp
+FF_VISITABLE_STRUCT(TownPopulation, people, cows); // ERROR: TownPopulation should not be default constructible
+```
+we get the something approximating the error in the comment.
+If we were to abandon `visitable` entirely, we would have to write
+```cpp
+struct Cow { ... };
+
+struct TownPopulation {
+  TownPopulation() = default;
+  TownPopulation(std::vector<Person> const &people,
+                 std::vector<Cow> const &cows)
+    : people(people), 
+      cows(cows)
+  { }
+
+  friend bool operator==(TownPopulation const &lhs, TownPopulation const &rhs) {
+    return lhs.people == rhs.people 
+      && lhs.cows == rhs.cows;
+  }
+
+  friend bool operator!=(TownPopulation const &lhs, TownPopulation const &rhs) {
+    return lhs.people != rhs.people
+      || lhs.cows != rhs.cows;
+  }
+
+  friend bool operator<(TownPopulation const &lhs, TownPopulation const &rhs) {
+    return lhs.people < rhs.people
+      || lhs.cows < rhs.cows;
+  }
+
+  std::vector<Person> people;
+  std::vector<Cow> cows;
+};
+
+namespace std {
+
+template <>
+struct hash<::TownPopulation> {
+  size_t operator()(::TownPopulation const &t) const {
+    size_t result = 0;
+    hash_combine(result, t.people);
+    hash_combine(result, t.cows);
+    return result;
+  }
+};
+
+}
+```
+which is tedious and bug-prone.
```

Comment:
this is a pretty long example just to say its tedious, that said it gets the point across if you cant think of another way to do so